powerstore/resource_smb_share.go

```diff
@@ -0,0 +1,658 @@
+		OfflineAvailability:             types.StringValue(shareDetails.OfflineAvailability),
+	}
+	smbShareACLAttrTypes := map[string]attr.Type{
+		"trustee_type": types.StringType,
+		"trustee_name": types.StringType,
+		"access_level": types.StringType,
+		"access_type":  types.StringType,
+	}
+
+	SmbShareElemTypes := types.ObjectType{
+		AttrTypes: smbShareACLAttrTypes,
+	}
+
+	objectIPs := []attr.Value{}
+	var diags diag.Diagnostics
+	for _, ip := range aclDetails.Aces {
+		obj := map[string]attr.Value{
+			"trustee_type": types.StringValue(ip.TrusteeType),
+			"trustee_name": types.StringValue(ip.TrusteeName),
+			"access_level": types.StringValue(ip.AccessLevel),
+			"access_type":  types.StringValue(ip.AccessType),
+		}
+		objVal, dgs := types.ObjectValue(smbShareACLAttrTypes, obj)
+		diags = append(diags, dgs...)
+		objectIPs = append(objectIPs, objVal)
+	}
+	setVal, dgs := types.SetValue(SmbShareElemTypes, objectIPs)
+	diags = append(diags, dgs...)
+	smbShare.SMBShareACL = setVal
+	return smbShare, diags
+}
+
+// getSMBShareDetails - method to get smb share details
+func (r *resourceSMBShare) getSMBShareDetails(ctx context.Context, SMBShareID string) (*models.SMBShare, diag.Diagnostics) {
+	var diags diag.Diagnostics
+	// Get SMBShare Details using ID retrieved above
+	SMBShareResponse, err := r.client.PStoreClient.GetSMBShare(context.Background(), SMBShareID)
+	if err != nil {
+		diags.AddError(
+			"Error reading smb share details",
+			"Could not get smb share, unexpected error: "+err.Error(),
+		)
+		return nil, diags
+	}
+
+	// Get SMBShare Details ACL using ID retrieved above
+	SMBShareACLResponse, err := r.client.PStoreClient.GetSMBShareACL(context.Background(), SMBShareID)
+	if err != nil {
+		diags.AddError(
+			"Error reading smb share ACL details",
+			"Could not get smb share, unexpected error: "+err.Error(),
+		)
+		return nil, diags
+	}
+
+	// Update details to state
+	state, diags := r.SMBShareStateACL(SMBShareResponse, SMBShareACLResponse)
+	if diags.HasError() {
+		return nil, diags
+	}
+	return &state, diags
+
+}
+
+// updateSMBShareACL - method to update smb share ACL
+func (r *resourceSMBShare) updateSMBShareACL(ctx context.Context, resp *resource.CreateResponse, SMBShareID string, aclParams *gopowerstore.ModifySMBShareACL) diag.Diagnostics {
+	var diags diag.Diagnostics
+	var err error
+	_, err = r.client.PStoreClient.SetSMBShareACL(context.Background(), SMBShareID, aclParams)
+	if err != nil {
+		diags.AddError(
+			"Error creating smb share ACL",
+			"Could not create smb share ACL, unexpected error: "+err.Error(),
+		)
+		// Delete SMBShare by calling API to prevent from getting tainted
+		_, err := r.client.PStoreClient.DeleteSMBShare(context.Background(), SMBShareID)
+		if err != nil {
+			diags.AddError(
+				"Error deleting SMBShare",
+				"Could not delete SMBShareID "+SMBShareID+": "+err.Error(),
+			)
+			return diags
+		}
+		// delete has passed so we need to remove state and return SMBShare ACL update failure
```

Comment:
What is the failure about in this comment?

Reply:
So, it says that we have updated the state once already after the create is done, now in case of failure of ACL API call, we will run delete. Let's say that delete fails then the state should remain as is and apart from ACL other details are updated (also the reason for update of state twice) and this comment says that since the delete is successfully passed then so now we will remove the state updated in first step.